fe/fe-core/src/main/java/com/starrocks/planner/PaimonScanNode.java

```diff
@@ -0,0 +1,262 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.paimon.table.source.ReadBuilder;
+import org.apache.paimon.table.source.Split;
+import org.apache.paimon.utils.InstantiationUtil;
+
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.starrocks.thrift.TExplainLevel.VERBOSE;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class PaimonScanNode extends ScanNode {
+    private static final Logger LOG = LogManager.getLogger(PaimonScanNode.class);
+    private final AtomicLong partitionIdGen = new AtomicLong(0L);
+    private PaimonTable paimonTable;
+    private final HDFSScanNodePredicates scanNodePredicates = new HDFSScanNodePredicates();
+    private final List<TScanRangeLocations> scanRangeLocationsList = new ArrayList<>();
+    private CloudConfiguration cloudConfiguration = null;
+
+    public PaimonScanNode(PlanNodeId id, TupleDescriptor desc, String planNodeName) {
+        super(id, desc, planNodeName);
+        paimonTable = (PaimonTable) desc.getTable();
+        setupCloudCredential();
+    }
+
+    public HDFSScanNodePredicates getScanNodePredicates() {
+        return scanNodePredicates;
+    }
+
+    public PaimonTable getPaimonTable() {
+        return paimonTable;
+    }
+
+    private void setupCloudCredential() {
+        String catalog = paimonTable.getCatalogName();
+        if (catalog == null) {
+            return;
+        }
+        PaimonConnector connector = (PaimonConnector) GlobalStateMgr.getCurrentState().getConnectorMgr().
+                getConnector(catalog);
+        if (connector != null) {
+            cloudConfiguration = connector.getCloudConfiguration();
+        }
+    }
+
+    @Override
+    protected String debugString() {
+        MoreObjects.ToStringHelper helper = MoreObjects.toStringHelper(this);
+        helper.addValue(super.debugString());
+        helper.addValue("paimonTable=" + paimonTable.getName());
+        return helper.toString();
+    }
+
+    @Override
+    public List<TScanRangeLocations> getScanRangeLocations(long maxScanRangeLength) {
+        return scanRangeLocationsList;
+    }
+
+    public void setupScanRangeLocations(DescriptorTable descTbl) throws AnalysisException {
+        ReadBuilder readBuilder = paimonTable.getNativeTable().newReadBuilder();
+        // TODO(wangriyu): readBuilder::withFilter
+        List<Split> splits = readBuilder.newScan().plan().splits();
+        for (Split split : splits) {
+            addScanRangeLocations(split);
+        }
+    }
+
+    private void addScanRangeLocations(Split split) {
+        TScanRangeLocations scanRangeLocations = new TScanRangeLocations();
+
+        THdfsScanRange hdfsScanRange = new THdfsScanRange();
+        hdfsScanRange.setUse_paimon_jni_reader(true);
+        hdfsScanRange.setPaimon_split_info(encodeObjectToString(split));
```

Comment:
Which split types do we currently support?

Reply:
all